core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,46 +197,65 @@ public void setStoreListenerSource(Cache<K, V> source) {
   }
 
   @Override
-  public void onEvent(final CacheEvent<K, V> event) {
+  public void onEvent(CacheEvent<K, V> event) {
     final EventType type = event.getType();
+    EventThreadLocal threadLocal = eventThreadLocal;
     LOGGER.trace("Cache Event notified for event type {}", type);
-    Map<EventListenerWrapper, Future<?>> notificationResults = 
-        new HashMap<CacheEventNotificationServiceImpl.EventListenerWrapper, Future<?>>(registeredListeners.size());
-    
-    for (final EventListenerWrapper wrapper: registeredListeners) {
-      if (!wrapper.config.fireOn().contains(type)) {
-        continue;
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    if (hasListeners()) {
+      threadLocal.verifyOrderedDispatch(orderedListenerCount.get() != 0);
+      threadLocal.addToEventList(cacheEventWrapper);
+      if (threadLocal.isOrdered()) {
+        ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue(event.getKey());
+        eventWrapperQueue.add(cacheEventWrapper);
       }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
-        }
-      };
-      
-      ExecutorService eventDelivery = wrapper.config.orderingMode().equals(EventOrdering.UNORDERED) ? unorderedDelivery : orderedDelivery;
-      notificationResults.put(wrapper, eventDelivery.submit(notificationTask));
-    }
-    
-    for (Map.Entry<EventListenerWrapper, Future<?>> entry: notificationResults.entrySet()) {
-      EventListenerWrapper wrapper = entry.getKey();
-      Future<?> f = entry.getValue();
-      if ((EventFiring.SYNCHRONOUS.equals(wrapper.config.firingMode()))) {
-        boolean interrupted = false;
-        try {
-          f.get();
-        } catch (ExecutionException e) {
-            LOGGER.error("Cache Event Listener failed for event type {} due to ", type, e);
-          // XXX delegate to resilience strategy (#52), and/or just log
-        } catch (InterruptedException e) {
-          interrupted = true;
-        } finally {
-          if (interrupted) {
-            Thread.currentThread().interrupt();
-          }
+    }
+  }
+
+  private void dispatchEvent(CacheEventWrapper cacheEventWrapper, EventThreadLocal currentEventThreadLocal) {
```

Comment:
Objective – receives the event that needs to be dispatched, processes them based on ordering and trigger dispatch

What it internally does:
•	Mark the event `FIREABLE`
•	In case of ordered event dispatch
o	Get the concerned queue that carries the event
o	Initiate wait if the event is not at the front of the queue
o	If event is at the front of the queue then for all events that are `FIREABLE`, 
•	Get the event from the front of queue
•	Send it for `actualDispatch()`
•	If event doesn’t belongs to the thread, mark it FIRED and signal FIRED-CONDITION
•	Else if it does belong to the thread, just mark the event FIRED
•	Remove the event from the event queue
•	Repeat until eventQueue empty or event at the front of the queue is not `FIREABLE` yet.
•	Send event for `actualDispatch()`

Why it should work:
The thread that created the event in the front of the queue is responsible for firing it. Others wait for their event to be fired (waiting on the FIRED_CONDITION), no matter sync or async. This helps to ensure that ordering is maintained and events that are `FIREABLE` can be processed as soon as the front of the concerned queue is `FIREABLE`. The approach of letting the thread with the event on the front of the queue fire events is preferred because having the approach of firing the event at the front and letting the next event know it’s the front now can cause conflicts and a requirement for additional conditions.